### ISSUE TITLE
Consider using subquery pushdown planner for non-subqueries as well

### ISSUE THREAD

**Issue by onderkalaci** [citusdata/citus#1706]
Recently I’ve had some discussions with different people on using subquery pushdown planner for non-subqueries as well. In this issue, I’d like to share my understanding and discuss the trade-offs for a broader discussion in the future.

Very briefly, subquery pushdown planner checks the query and decides whether a multi-shard query is safe to pushdown (almost) as is. If the answer is yes, the query is slightly changed (for aggregates on the top-level) and sent to the workers via real-time executor. Finally, aggregations on the coordinator is done and the result is returned to the users. ([Here](https://docs.google.com/document/d/1azuhKuzkw7cMJJ6srQNjvugbiBfpP01EzjHiWPSTolU/edit) is the complete details of the planner.)

I'll refer the planner with the name `MultiPushdown` planner, which is basically the existing subquery pushdown planner (and name suggestions are welcome). 

On the high-level, what I had in mind is the following: Citus planner gets the query. It first checks whether the query is router plannable or not. If not, Citus tries `MultiPushdown` planner. If that fails too, Citus tries the existing multi-shard planner.

(Here is a simple [sketch](https://docs.google.com/drawings/d/1FbVfqgWuVlz_F-GbA8Od4a_xmimsxLKE0vi8hhjhzrs/edit) of the proposed planning logic)

There are two important things to note about `MultiPushdown` planner:

* It can already be used for non-subqueries. For instance, INSERT/SELECT pushdown logic uses the exact functionality with subquery pushdown even for non-subqueries such as [here](https://github.com/citusdata/citus/blob/master/src/test/regress/sql/multi_insert_select.sql#L51).
* It isn’t a completely separate planner. Instead, the output of the planner is represented via regular nodes in the multi-relational algebra (i.e., `MultiCollect` -> `MultiTable`).   
    * Note that this is crucial since it enables us to run a wide range of queries including non-partition key `GROUP BY`s on the top level query. (In non-subquery case, the top level query is the input query itself!) 
    * This implies that a good chunk of real-time queries that today Citus supports could be planned via this planner.

### The motivation for the proposal:

* Wider SQL coverage for non-subqueries
    * Non-equi reference table joins
    * Pushdownable window functions
    * Pushdownable UNION/UNION ALL (We need to improve this first, the coverage is low now)
    * (There used to be DISTINCT/DISTINCT ON, but we've already implemented those)
* Remove FixOrderJoinPlanner (i.e., Outer join planner) at all
    * See the discussions with append distributed tables below
    * Could potentially remove some parts of LocalJoin code-path?
* Eliminate the special treatment of subqueries in the planner. As we increase coverage subqueries, non-subqueries would benefit from that as well.
* The query planning times of relatively simpler queries wouldn’t be affected by the future changes on the logical planner/logical optimizer.
* Simpler to communicate. Unifies the planning which does the same thing. (INSERT/SELECT pushdown planning and subquery pushdown planning both of which utilize router planning at the core.)

### Cons of the approach:
* Currently the `MultiPushdown` planner doesn’t work with append distributed tables. We could potentially enable it, but, this requires CPU/memory intensive co-location and non-overlapping shard checks.
    * Still, we have to reject non-coloated tables in the query and tables with overlapping shards
    * We should probably just not implement that because it doesn't make much sense.
    * But, if we don’t implement this for append distributed tables, then we cannot remove any code (i.e., outer join order planning) or we don’t support those features for append distributed tables (which sounds disappointing).

* There are two open issues on the performance of planning [#1332, #1245]. 
    * The algorithmic complexity is high, but as benchmarked in #1332 it is not actually becoming a bottleneck (thanks to latest changes on the shard pruning). 
   * I'd expect the planning times to increase if we don't fix the above issues. But, I don't expect that it'd be noticeable.
    * We should simply pick some representative queries and benchmark the effect of the change before deciding to fix them.
* We have a limited UNION/UNION ALL support compared to what we can actually support. This might give a bad user experience. We should consider fixing it before this one.
* We're touching a large customer facing feature. Requires extensive testing & benchmarking etc.
* We're handling prepared statements slightly differently for subquery pushdown planning (since it relies on the `unmodifiedQuery` as opposed to the query that passed through `standard_planner`)

### Things to do (very roughly):

	Firstly, I should mention that this task is far from being a trivial task.

* Refactor planner 
    * So that it flows Router -> Multi Pushdown -> Existing Logical Planning
    * All error checks should be DeferredError
    * Refactor the name `subquery` to `MultiPushdown` (or smth else) in the code / comments / planner structs. Explicitly represent the pushdown node in the plan tree (`MultiTable` doesn’t sound the best name/representation). Get rid of all tricks around subquery pushdown planning (i.e., `SUBQUERY_PUSHDOWN_RELATION_ID`)
* Remove FixedJoinOrderPlanner
* We have too many tests on append distributed tables that this change touches. And, we don't seem to have their equivalents for hash distributed tables.
    * Add the equivalent number of hash distributed regression tests as that we have for append
    * Many of the TPCH queries that we support start to fail with the default shard count for hash distributed tables since we rely on broadcast joins by setting the `citus.large_table_shard_count`
    * We have a fair number of tests that rely on `citus.large_table_shard_count` to prevent re-partitioning. We should not skip adding any feature tests for hash distributed tables.

To sum up, although the benefits are significant, due to the interactions with append distribution and complexity of the implementation & testing, I’m doubtful of implementing this soon. If one day we decide to drop the support for append distribution, we could consider implementing this. But I still find opening the issue useful. 

Also, if we’re to consider implementing this feature, I could update the specifics of the issue, 
 specifically propose a document for refactoring the planner. 

[Here](https://github.com/citusdata/citus/compare/multi_pushdown_planner?expand=1) is a very very WIP prototype of it.

**Comment by marcocitus** [comment]
This seems like it would be a huge improved in both code structure and usability.

Is there a way we could cover cases like LIMIT or top-level aggregates? It seems that if we want to do all outer joins through this logic then we would need to cover whatever the logical planner supports today.

**Comment by onderkalaci** [comment]
> It seems that if we want to do all outer joins through this logic then we would need to cover whatever the logical planner supports today.

That's already the case, I tried to explain in the description in detail. Here is the related part:

* It isn’t a completely separate planner. Instead, the output of the planner is represented via regular nodes in the multi-relational algebra (i.e., MultiCollect -> MultiTable).
    * Note that this is crucial since it enables us to run a wide range of queries including non-partition key GROUP BYs on the top level query. (In non-subquery case, the top level query is the input query itself!)
    * This implies that a good chunk of real-time queries that today Citus supports could be planned via this planner.

**Comment by onderkalaci** [comment]
> Is there a way we could cover cases like LIMIT o

And, we already allow LIMITs in the top level query where for non-subquery case the top-level query is the input query.

**Comment by marcocitus** [comment]
Cool, just wanted to make sure that a query like `SELECT * FROM a LEFT JOIN b USING (id) LIMIT 10` is still supported and that the limit is pushed down.

**Comment by marcocitus** [comment]
Implemented
